### PR TITLE
expression: rewrite builtin function: NOW, UTC_TIMESTAMP, UTC_DATE

### PR DESCRIPTION
**Pull request by zz-jason**
to https://github.com/pingcap/tidb/issues/4080

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
ci failed @zz-jason

**Comment by zz-jason** [comment]
@XuHuaiyu  fixed, PTAL

**Comment by zz-jason** [comment]
@XuHuaiyu @breeswish @jackysp PTAL

**Comment by breeswish** [comment]
LGTM

File: expression/builtin_time.go [hunk 1]
```diff
@@ -704,59 +706,104 @@ type nowFunctionClass struct {
...
-		return d, errors.Trace(err)
+		return types.Time{}, true, errors.Trace(err)
+	}
+	if !isNull && num > int64(types.MaxFsp) {
+		return types.Time{}, true, errors.Errorf("Too-big precision %v specified for 'now'. Maximum is %v.", num, types.MaxFsp)
+	}
+	if !isNull && num < int64(types.MinFsp) {
+		return types.Time{}, true, errors.Errorf("Invalid negative %d specified, must in [0, 6].", num)
+	}
+	if isNull {
```
Comment (hanfei1991) [inline expression/builtin_time.go]: These three checks can write as `if isNull {} else if num > .. else if num < ...{}`

File: expression/builtin_time.go [hunk 2]
```diff
@@ -1505,64 +1552,121 @@ func (c *utcDateFunctionClass) getFunction(args []Expression, ctx context.Contex
...
+		Fsp:  types.UnspecifiedFsp}
+	return result, false, nil
 }
 
 type utcTimestampFunctionClass struct {
 	baseFunctionClass
 }
 
+func (c *utcTimestampFunctionClass) getFlenAndDecimal(sc *variable.StatementContext, arg Expression) (flen, decimal int) {
+	if constant, ok := arg.(*Constant); ok {
```
Comment (hanfei1991) [inline expression/builtin_time.go]: It seems it's same as `now`'s getFlenAndDecimal. Can we reuse ?

File: expression/builtin_time.go (hunk 1 above)
Comment (zz-jason) [inline expression/builtin_time.go]: done

File: expression/builtin_time.go (hunk 2 above)
Comment (zz-jason) [inline expression/builtin_time.go]: done

File: expression/builtin_time.go [hunk 3]
```diff
@@ -1669,6 +1565,112 @@ func (b *builtinUTCTimestampWithoutArgSig) evalTime(row []types.Datum) (types.Ti
 	return result, isNull, errors.Trace(err)
 }
 
+type nowFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *nowFunctionClass) getFlenAndDecimal(sc *variable.StatementContext, arg Expression) (flen, decimal int) {
```
Comment (hanfei1991) [inline expression/builtin_time.go]: Can this func be removed?

**Pull request merged**